File: benchmarks/runtime/kotoba_bench/scalar.clj
(ns kotoba-bench.scalar)

(set! *warn-on-reflection* true)

(defn multiply [left right]
  (unchecked-multiply (long left) (long right)))

(defn checksum [iterations]
  (let [limit (long iterations)
        product (long (multiply 6 7))]
    (loop [index (long 0) total (long 0)]
      (if (< index limit)
        (recur (unchecked-inc index) (unchecked-add total product))
        total))))

(defn branch-checksum [iterations]
  (let [limit (long iterations)
        half (quot limit 2)]
    (loop [index (long 0) total (long 0)]
      (if (< index limit)
        (recur (unchecked-inc index)
               (unchecked-add total (if (< index half) (long 41) (long 43))))
        total))))

(defn xorshift32-checksum [iterations]
  (let [limit (long iterations)
        mask 0xffffffff]
    (loop [index (long 0) state (long 2463534242)]
      (if (< index limit)
        (let [x1 (bit-and mask (bit-xor state (bit-shift-left state 13)))
              x2 (bit-and mask (bit-xor x1 (unsigned-bit-shift-right x1 17)))
              x3 (bit-and mask (bit-xor x2 (bit-shift-left x2 5)))]
          (recur (unchecked-inc index) x3))
        state))))

(defn vector-allocation-checksum [iterations]
  (let [limit (long iterations)
        mask (long 0xffffffff)]
    (loop [index (long 0) state (long 2463534242) total (long 0)]
      (if (< index limit)
        (let [values (vector 3 5 8 13 21 34 55 89)
              x1 (bit-and mask (bit-xor state (bit-shift-left state 13)))
              x2 (bit-and mask (bit-xor x1 (unsigned-bit-shift-right x1 17)))
              next (bit-and mask (bit-xor x2 (bit-shift-left x2 5)))]
          (recur (unchecked-inc index) next
                 (unchecked-add total (long (nth values (bit-and next 7))))))
        total))))

(defn retain-vector [values]
  values)

(defn vector-materialization-checksum [iterations]
  (let [limit (long iterations)
        mask (long 0xffffffff)]
    (loop [index (long 0) state (long 2463534242) total (long 0)]
      (if (< index limit)
        (let [values (vector 3 5 8 13 21 34 55 89)
              x1 (bit-and mask (bit-xor state (bit-shift-left state 13)))
              x2 (bit-and mask (bit-xor x1 (unsigned-bit-shift-right x1 17)))
              next (bit-and mask (bit-xor x2 (bit-shift-left x2 5)))]
          (recur (unchecked-inc index) next
                 (unchecked-add total
                                (long (nth (if (zero? (bit-and index 511))
                                             (retain-vector values)
                                             values)
                                           (bit-and next 7))))))
        total))))

(defn -main [& args]
  (let [workload (or (first args) "scalar")
        once? (= "--once" (second args))
        iterations (Long/parseLong (or (nth args 2 nil) "1"))
        warmup-iterations (Long/parseLong
                           (or (nth args 3 nil) (str (min iterations 1000))))
        run-workload (case workload
                       "scalar" checksum
                       "branch" branch-checksum
                       "mix" xorshift32-checksum
                       "vector" vector-allocation-checksum
                       "vector-materialize" vector-materialization-checksum
                       (throw (ex-info "unknown benchmark workload" {:workload workload})))]
    (if once?
      (println (str "{\"checksum\":" (run-workload iterations) "}"))
      (do
        (run-workload warmup-iterations)
        (let [started (System/nanoTime)
              result (run-workload iterations)
              elapsed (- (System/nanoTime) started)]
          (println (str "{\"iterations\":" iterations
                        ",\"warmupIterations\":" warmup-iterations
                        ",\"checksum\":" result
                        ",\"elapsedNanoseconds\":" elapsed "}")))))))

File: benchmarks/runtime/scalar.rs
use std::env;
use std::hint::black_box;
use std::time::Instant;

#[inline(never)]
fn multiply(left: i64, right: i64) -> i64 {
    black_box(left).wrapping_mul(black_box(right))
}

fn checksum(iterations: u64) -> i64 {
    let mut total = 0_i64;
    for _ in 0..iterations {
        total = black_box(total.wrapping_add(multiply(6, 7)));
    }
    total
}

fn branch_checksum(iterations: u64) -> i64 {
    let half = iterations / 2;
    let mut total = 0_i64;
    for index in 0..iterations {
        total = black_box(total.wrapping_add(if black_box(index) < half { 41 } else { 43 }));
    }
    total
}

fn xorshift32_checksum(iterations: u64) -> i64 {
    let mut state = 2_463_534_242_u32;
    for _ in 0..iterations {
        state ^= state << 13;
        state ^= state >> 17;
        state ^= state << 5;
    }
    i64::from(state)
}

fn vector_allocation_checksum(iterations: u64) -> i64 {
    let mut state = 2_463_534_242_u32;
    let mut total = 0_i64;
    for _ in 0..iterations {
        let values = black_box([3_i64, 5, 8, 13, 21, 34, 55, 89]);
        state ^= state << 13;
        state ^= state >> 17;
        state ^= state << 5;
        total = total.wrapping_add(values[(state & 7) as usize]);
    }
    total
}

#[inline(never)]
fn retain_vector(values: [i64; 8]) -> [i64; 8] {
    black_box(values)
}

fn vector_materialization_checksum(iterations: u64) -> i64 {
    let mut state = 2_463_534_242_u32;
    let mut total = 0_i64;
    for index in 0..iterations {
        let values = black_box([3_i64, 5, 8, 13, 21, 34, 55, 89]);
        state ^= state << 13;
        state ^= state >> 17;
        state ^= state << 5;
        let selected = if index & 511 == 0 {
            retain_vector(values)[(state & 7) as usize]
        } else {
            values[(state & 7) as usize]
        };
        total = total.wrapping_add(selected);
    }
    total
}

fn main() {
    let args: Vec<String> = env::args().skip(1).collect();
    let workload = args.first().map(String::as_str).unwrap_or("scalar");
    let once = args.get(1).map(String::as_str) == Some("--once");
    let iterations = args.get(2).and_then(|value| value.parse().ok()).unwrap_or(1);
    let warmup_iterations = args
        .get(3)
        .and_then(|value| value.parse().ok())
        .unwrap_or(iterations.min(1_000));
    let run = |count| match workload {
        "scalar" => checksum(count),
        "branch" => branch_checksum(count),
        "mix" => xorshift32_checksum(count),
        "vector" => vector_allocation_checksum(count),
        "vector-materialize" => vector_materialization_checksum(count),
        _ => panic!("unknown benchmark workload"),
    };
    if once {
        println!("{{\"checksum\":{}}}", run(iterations));
        return;
    }
    black_box(run(warmup_iterations));
    let started = Instant::now();
    let result = black_box(run(iterations));
    let elapsed = started.elapsed().as_nanos();
    println!(
        "{{\"iterations\":{},\"warmupIterations\":{},\"checksum\":{},\"elapsedNanoseconds\":{}}}",
        iterations, warmup_iterations, result, elapsed
    );
}
